File: Dynamic Programming/Python/Solving Questions With Brainpower.py
# Problem link: https://leetcode.com/problems/solving-questions-with-brainpower/description/
# Time: O(n)
# Space: O(n)
class Solution:
    def mostPoints(self, questions: List[List[int]]) -> int:
        DP = [0] * len(questions)

        for index in range(len(questions) - 1, -1, -1):
            points, brainPower = questions[index]
            withQ = withoutQ = 0

            if index + brainPower + 1 < len(questions):
                withQ = DP[index + brainPower + 1]
            if index + 1 < len(questions):
                withoutQ = DP[index + 1]
            
            DP[index] = max(points + withQ, withoutQ)
        
        return DP[0]

File: Dynamic Programming/Java/Solving Questions With Brainpower.java
// Problem link: https://leetcode.com/problems/solving-questions-with-brainpower/description/
// Time: O(n)
// Space: O(n)
class Solution {
    public long mostPoints(int[][] questions) {
        int size = questions.length, points, brainPower;
        long withQ, withoutQ, DP[] = new long[size];

        for(int index = size - 1; index >= 0; index--) {
            points = questions[index][0];
            brainPower = questions[index][1];
            withQ = withoutQ = 0;

            if(index + brainPower + 1 < size) {
                withQ = DP[index + brainPower + 1];
            }
            if(index + 1 < size) {
                withoutQ = DP[index + 1];
            }

            DP[index] = Math.max(points + withQ, withoutQ);
        }


        return DP[0];
    }
}
